linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala

```diff
@@ -375,112 +375,152 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
-            lookupClassForLinking(_, origin))
-      } yield {
-        val maybeCycle = maybeAncestors.collectFirst {
-          case cycle @ CycleInfo(_, null) => cycle
-
-          case CycleInfo(c, root) if root == className =>
-            CycleInfo(className :: c, null)
-
-          case CycleInfo(c, root) =>
-            CycleInfo(className :: c, root)
+      // Start loading the parents.
+      val parentNames = data.superClass.toList ++ data.interfaces
+      val parentsFuture = Future.traverse(parentNames) {
+        ensureLoading(_, syntheticKind = None) match {
+          case l: LoadingInfos    => l.result
+          case l: InfoLoadedState => Future.successful(l)
         }
+      }
 
-        maybeCycle.getOrElse {
-          val ancestors = maybeAncestors.asInstanceOf[List[ClassInfo]]
-
-          val (superClass, interfaces) =
-            if (data.superClass.isEmpty) (None, ancestors)
-            else (Some(ancestors.head), ancestors.tail)
+      // Register that we have completed info loading.
+      // We must do concurrently with parent loading. Otherwise, we'll block
+      // indefinitely if there is a cycle.
+      val linkingClass = new LinkingClass(className, parentsFuture)
+      _classInfos.put(className, linkingClass)
+      loadingInfos.complete(linkingClass)
+
+      linkingClass.checkParentChain().flatMap {
+        case Some(cyclePath) =>
+          _errors ::= CycleInInheritanceChain(cyclePath, fromAnalyzer)
+          Future.successful(InheritanceCycle)
+
+        case None =>
+          // There is no cycle. We can safely wait until all parents are loaded.
+          parentsFuture.flatMap { parents =>
+            Future.traverse(parents) {
+              case l: LinkingClass  => l.result
+              case r: LoadingResult => Future.successful(r)
+            }
+          }.map { loadedParents0 =>
+            val loadedParents = loadedParents0.asInstanceOf[List[ClassInfo]]
 
-          val info = new ClassInfo(data, superClass, interfaces, syntheticKind, nonExistent)
+            val (superClass, interfaces) =
+              if (data.superClass.isEmpty) (None, loadedParents)
+              else (Some(loadedParents.head), loadedParents.tail)
 
-          _classInfos.put(className, info)
+            val info = new ClassInfo(data, superClass, interfaces, syntheticKind, nonExistent)
 
-          implicit val from = FromClass(info)
-          ancestors.foreach(_.link())
+            implicit val from = FromClass(info)
+            loadedParents.foreach(_.link())
 
-          info
-        }
+            info
+          }
+      }.foreach { result =>
+        _classInfos.put(className, result)
+        linkingClass.complete(result)
       }
     }
   }
 
-  private sealed trait LoadingResult
   private sealed trait ClassLoadingState
+  private sealed trait InfoLoadedState extends ClassLoadingState
+  private sealed trait LoadingResult extends InfoLoadedState
```

Comment:
Perhaps consider adding methods on these types to resolve them to the possible tiers. IIUC the tiers we need are
```scala
  private sealed trait ClassLoadingState {
    def resolveToInfoLoadedState: Future[InfoLoadedState]
    def resolveToLoadingResult: Future[LoadingResult]
  }
```
This would lighten the notation inside the `traverse` calls. More importantly, IMO it would ease understanding that all cases are really covered.

(They could be implemented either in an OO fashion or with pattern matching.)

Reply:
Thanks. I find it much better this way.